File: ios/Classes/ClientApi/ClientApiMethodCallHandler.swift
/*
 *
 *  * Do not remove or alter the notices in this preamble.
 *  *
 *  * This software is owned by Worldline Global Collect and may not be be altered, copied, reproduced, republished, uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *  *
 *  * Copyright © 2023 Worldline Global Collect and/or its affiliates.
 *  *
 *  * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *  *
 *  * Please contact Worldline Global Collect for questions regarding license and user rights.
 *
 */

import Flutter
import WorldlineConnectKit

internal class ClientApiMethodCallHandler: ConnectSdkPlugin {
    private let connectSDKBridge = ConnectSDKBridge.shared

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPublicKey":
            getPublicKey(result)
        case "getBasicPaymentProducts":
            getBasicPaymentProducts(result)
        case "getPaymentProduct":
            getPaymentProduct(call, result)
        case "getPrivacyPolicy":
            getPrivacyPolicy(call, result)
        case "getIinDetails":
            getIinDetails(call, result)
        case "getBasicPaymentProductGroups":
            getBasicPaymentProductGroups(result)
        case "getPaymentProductGroup":
            getPaymentProductGroup(call, result)
        case "getPaymentItems":
            getPaymentItems(result)
        case "getPaymentProductDirectory":
            getPaymentProductDirectory(call, result)
        case "getThirdPartyStatus":
            getThirdPartyStatus(call, result)
        case "convertAmount":
            convertAmount(call, result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getPublicKey(_ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            self.connectSDKBridge.getPublicKey(result: result)
        }
    }
    
    private func getBasicPaymentProducts(_ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            self.connectSDKBridge.getBasicPaymentProducts(result: result)
        }
    }

    private func getPaymentProduct(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let paymentProductRequest = PaymentProductRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getPaymentProduct(request: paymentProductRequest, result: result)
        }
    }
    
    private func getPrivacyPolicy(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let privacyPolicyRequest = PrivacyPolicyRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getPrivacyPolicy(request: privacyPolicyRequest, result: result)
        }
    }

    private func getIinDetails(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let iinDetailsRequest = IinDetailsRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getIinDetails(request: iinDetailsRequest, result: result)
        }
    }

    private func getBasicPaymentProductGroups(_ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            self.connectSDKBridge.getPaymentProductGroups(result: result)
        }
    }
    
    private func getPaymentProductGroup(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let paymentProductGroupRequest = PaymentProductGroupRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getPaymentProductGroup(request: paymentProductGroupRequest, result: result)
        }
    }
    
    private func getPaymentItems(_ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            self.connectSDKBridge.getPaymentItems(result: result)
        }
    }
    
    private func getPaymentProductDirectory(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let paymentProductDirectoryRequest = PaymentProductDirectoryRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getPaymentProductDirectory(request: paymentProductDirectoryRequest, result: result)
        }
    }
    
    private func getThirdPartyStatus(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let thirdPartyStatusRequest = ThirdPartyStatusRequest.getRequestObject(for: call, result: result) else {
                return
            }
            
            self.connectSDKBridge.getThirdPartyStatus(request: thirdPartyStatusRequest, result: result)
        }
    }
    
    private func convertAmount(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        connectSDKBridge.executeWhenInitialized(result) {
            guard let convertAmountRequest = ConvertAmountRequest.getRequestObject(for: call, result: result) else {
                return
            }

            self.connectSDKBridge.convertAmount(request: convertAmountRequest, result: result)
        }
    }
}

File: android/src/main/kotlin/com/worldline/connect/connect_sdk_client_flutter/client_api/ClientApiMethodCallHandler.kt
/*
 *
 *  * Do not remove or alter the notices in this preamble.
 *  *
 *  * This software is owned by Worldline Global Collect and may not be be altered, copied, reproduced, republished, uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *  *
 *  * Copyright © 2023 Worldline Global Collect and/or its affiliates.
 *  *
 *  * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *  *
 *  * Please contact Worldline Global Collect for questions regarding license and user rights.
 *
 */

package com.worldline.connect.connect_sdk_client_flutter.client_api

import com.google.gson.Gson
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.ConvertAmountRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.PrivacyPolicyRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.IinDetailsRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.PaymentProductDirectoryRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.PaymentProductGroupRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.PaymentProductRequest
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.ResultError
import com.worldline.connect.connect_sdk_client_flutter.client_api.models.ThirdPartyStatusRequest
import com.worldline.connect.connect_sdk_client_flutter.connect_sdk.ConnectSDKBridge
import com.worldline.connect.connect_sdk_client_flutter.getRequestOrReturnError
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class ClientApiMethodCallHandler(private val gson: Gson)
    : MethodChannel.MethodCallHandler {
    companion object {
        private const val TAG = "ApiCallMethodHandler"
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPublicKey" -> getPublicKey(result)
            "getBasicPaymentProducts" -> getBasicPaymentProducts(result)
            "getPaymentProduct" -> getPaymentProduct(call, result)
            "getPrivacyPolicy" -> getPrivacyPolicy(call, result)
            "getIinDetails" -> getIinDetails(call, result)
            "getBasicPaymentProductGroups" -> getBasicPaymentProductGroups(result)
            "getPaymentProductGroup" -> getPaymentProductGroup(call, result)
            "getPaymentItems" -> getPaymentItems(result)
            "getPaymentProductDirectory" -> getPaymentProductDirectory(call, result)
            "getThirdPartyStatus" -> getThirdPartyStatus(call, result)
            "convertAmount" -> convertAmount(call, result)
            else -> result.notImplemented()
        }
    }

    private fun callSdkIfInitialized(result: MethodChannel.Result, runWhenInitialized: () -> Unit) {
        if (ConnectSDKBridge.isInitialized()) {
            runWhenInitialized()
        } else {
            return ResultError.sdkNotInitialized(result)
        }
    }

    private fun getPublicKey(result: MethodChannel.Result) {
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPublicKey(result)
        }
    }

    private fun getBasicPaymentProducts(result: MethodChannel.Result) {
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getBasicPaymentProducts(result)
        }
    }

    private fun getPaymentProduct(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(PaymentProductRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPaymentProduct(result, request)
        }
    }

    private fun getPrivacyPolicy(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(PrivacyPolicyRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPrivacyPolicy(result, request)
        }
    }

    private fun getIinDetails(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(IinDetailsRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getIinDetails(result, request)
        }
    }

    private fun getBasicPaymentProductGroups(result: MethodChannel.Result) {
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPaymentProductGroups(result)
        }
    }

    private fun getPaymentProductGroup(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(PaymentProductGroupRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPaymentProductGroup(result, request)
        }
    }

    private fun getPaymentItems(result: MethodChannel.Result) {
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPaymentItems(result)
        }
    }

    private fun getPaymentProductDirectory(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(PaymentProductDirectoryRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getPaymentProductDirectory(result, request)
        }
    }

    private fun getThirdPartyStatus(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(ThirdPartyStatusRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.getThirdPartyStatus(result, request)
        }
    }

    private fun convertAmount(call: MethodCall, result: MethodChannel.Result) {
        val request = call.getRequestOrReturnError(ConvertAmountRequest::class.java, result, gson)
        request ?: return
        callSdkIfInitialized(result) {
            ConnectSDKBridge.convertAmount(result, request)
        }
    }
}